src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter/ControllerSaveTempDataPropertyFilterFactory.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Reflection;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public class ControllerSaveTempDataPropertyFilterFactory : IFilterFactory
+    {
+        public IList<TempDataProperty> TempDataProperties { get; set; }
+
+        public bool IsReusable => false;
+
+        public IFilterMetadata CreateInstance(IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            var service = serviceProvider.GetRequiredService<ControllerSaveTempDataPropertyFilter>();
+            service.TempDataProperties = TempDataProperties;
+            return service;
+        }
+
+        public static IList<TempDataProperty> GetTempDataPropertyHelpers(Type modelType)
```

Comment:
This should probably be on `SaveTempDataPropertyFilterBase` if it's going to be public.